Core/clim-core/commands/tables.lisp

```diff
@@ -326,54 +326,54 @@ designator) inherits menu items."
-    (values)))
-
-(defun find-menu-item (menu-name command-table &key (errorp t))
-  (loop with table = (find-command-table command-table)
-        for item in (slot-value table 'menu)
-        when (string-equal menu-name (command-menu-item-name item))
-          do (return-from find-menu-item
-               (values item command-table)))
-  (when (inherit-menu-items command-table)
-    (loop for table in (command-table-inherit-from command-table)
-          do (multiple-value-bind (found-item source-table)
-                 (find-menu-item menu-name table :errorp nil)
-               (when found-item
-                 (return-from find-menu-item
-                   (values found-item source-table))))))
+;;; This internal function is like map-over-command-menu-items, but it maps
+;;; over each element in the command table and (if inherited is t) its
+;;; ancestors. It does not descend into sub-menus. The mapped function accepts
+;;; two arguments: the item and the command table the item belongs to.
+(defun map-over-command-table-menu (function table &key (inherited t))
+  (setf table (find-command-table table))
+  (flet ((map-table-entries (table)
+           (mapc #'(lambda (item)
+                     (with-slots (menu-name keystroke type) item
+                       (funcall function item table)))
+                 (slot-value table 'menu))))
+    (map-table-entries table)
+    (when inherited
+      (dolist (sub-table (command-table-inherit-from table))
+        (map-over-command-table-menu function sub-table)))))
+
+(defun map-over-command-table-menu-items (function table &key (inherited t))
+  "Applies function to all of the items in `table's menu. `table' must be a
+command table or the name of a command table. `Function' must be a function of
+three arguments, the menu name (which will be NIL if there is none), the
+keystroke accelerator gesture (which will be NIL if there is none), and the
+command menu item; it has dynamic extent. The command menu items are mapped
+over in the order specified by `add-menu-item-to-command-table'. `table' is a
+command table designator. Any inherited menu items will be mapped over after
+`command-table's own menu items.
+
+`Map-over-command-table-menu-items' does not descend into sub-menus. If the
+programmer requires this behavior, he should examine the type of the command
+menu item to see if it is `:menu'."
+  (flet ((fun (item table)
+           (declare (ignore table))
+           (with-slots (menu-name keystroke type) item
+             (when (or menu-name (eq type :divider))
+               (funcall function menu-name keystroke item)))))
+    (map-over-command-table-menu
+     #'fun table :inherited (and inherited (inherit-menu-items table))))
+  (values))
+
+(defun find-menu-item (menu-name table &key (errorp t))
+  (flet ((fun (item table)
+           (when-let ((item-name (command-menu-item-name item)))
+             (when (string-equal item-name menu-name)
```

Comment:
Isn't this just `(when (equapl menu-name (command-menu-item-name item)) …` since `nil` is not `equalp` to any string?

Reply:
I think that I've made a mistake by using string-equal, I'm changing that calls to simply equal.

Reply:
But the `when-let` shouldn't be needed regardless since search for `nil` as name wouldn't have worked in the old version either.

But see above comment regarding case-insensitivity of menu item names.

Reply:
when-let is needed, because now we store both named and keystroke-only items in the menu. without when-let, call to (find-menu-item nil table) if there are keystroke-only entries would return such entry.

of course it is could be argued, that menu-name being nil is UB, or that we should that it is a string.

Reply:
Wouldn't validating `menu-name` be better, then?

Reply:
it already paid off, while testing the change I've found an issue somewhere else.

validating is fixed and the uncovered issue is fixed in a separate commit 

183aade7 remove-command-from-command-table: remove the matching menu entry

Reply:
Nice! Could you add a test which checks that the menu entry is removed? Should be easy with all the testing machinery you added.

Reply:
I've simply extended the test commands.add-and-remove - fixed.